### ISSUE TITLE
(s)Multiple user login is not allowed in mobile

### ISSUE THREAD

**Issue by sworupSussol** [openmsupply/mobile#760]
**Related to mSupply repo's issue:** https://github.com/sussol/msupply/issues/2191

Description
---
- We want multiple users to be able to log into the same Mobile tablet system. 
- We want the multi user functionality to work offline.
- *(In discussion)* We want same user to be able to log into multiple stores

----

Here's how the login system for mobile is working:

- Mobile does have a user table and the table is not limited to single user. It can handle multiple users.
- User is never a part of sync. 
- When the User first logs in the application authenticates with the server
- The server returns a json object on successful login which is saved in Realm database in mobile
- If a second user tries to log in, the process repeats. Authentication with the server and returned json being stored as second record of Realm database user table in mobile.
- In case of offline mode the application check against the realm database hence there has to be records there before going offline.
- Every time there is an authentication attempt the username entry is updated with latest password (manually simulated sync behaviour).

Conclusion
---

> Multi-user functionality actually is working in our Mobile system,  provided the user has logged in at least once with the server (online mode) and Realm has the records in its table.
- Since the User table is never synced, the user has to ***log in at least once*** with the server. 
- Once the user has logged in, his/her record gets created in Realm database, after which simultaneous offline login is not a problem.
- I have tried loggin in with different and/or same user from two different devices, online and offline mode and it works.

Solution
---

As @Chris-Petty said, we should include Users to sync and sync user table to Realm database when the initial sync happens. We should keep the User table in sync as any other table in the database.

**Comment by ujwalSussol** [comment]
@sworupSussol  : I think I understand the issue .....    I think we want to allow different staff to be able to login to the one tablet store.  ....  However I can see that the issue can be understood in different ways ... so will check with @Chris-Petty

**Comment by Chris-Petty** [comment]
@ujwalSussol that is more less right, just a couple caveats.
@sworupSussol I'm going to hijack your comment (so it's at the top) and add some detail if you don't mind.

**Comment by sworupSussol** [comment]
@Chris-Petty @ujwalSussol 

I did some research on this and here's how the login system for mobile is working:

- Mobile does have a user table and the table is not limited to single user. It can handle multiple users.
- User is never a part of sync. 
- When the User first logs in the application authenticates with the server
- The server returns a json object on successful login which is saved in Realm database in mobile
- If a second user tries to log in, the process repeats, authentication with the server and returned json is stored in second record of realm user database in mobile.
- In case of offline mode the application check against the realm database
- Every time there is an authentication attempt the username entry is updated with latest password

Conclusion
---

- Since the User table is never synced, the user has to log in at least once with the server. 
- Once the user has logged in once his/her record gets created in Realm database, after which simultaneous offline login is not a problem.
- I have tried loggin in with different and/or same user from two different devices, online and offline mode and it works if the user has logged in at least once with the server.

Solution
---

As @Chris-Petty said, we should include Users to sync and sync user table to Realm database when the initial sync happens. We should keep the User table in sync as any other table in the database. 

> I will work on this solution.

**Comment by ujwalSussol** [comment]
@Chris-Petty   : Like mSupply Desktop are we expecting to display the list of available users either as a user list or as a Drop down list.  If yes then we @sworupSussol  may need to work on a design first.

**Comment by Chris-Petty** [comment]
@ujwalSussol That's an interesting idea. If we want to limit the scope we can say nah, people should just remember their user name. Or maybe suggested auto complete as you type. Or a list as you say.

How about, @sworupSussol just run with this issue as making the User table included in sync for mobile. Make another issue talking for changing the interface to take advantage of the table once this issue is done (someone could even work on it in parallel, as the table is already defined).

**Comment by sworupSussol** [comment]
@Chris-Petty @andreievg

I have this all but done but I am stuck at one point.

The hashpassword for the user, when I tried to get the password field from 4D database it comes in as an encrypted text. I tried base64 decoding it but to no avail.

One solution could be convert the password from blob to hash text in 4D side, but that adds a lot of overhead. 

I can use a suggestion.

**Comment by andreievg** [comment]
@sworupSussol, i have a suggesting, and this would reduce the work load (it has a limitation though).
Create a user table, but instead of syncing the user tables, save users in that table when they have been successfully authenticated with the server (instead of saving it in the settings table).

The limitations of this method:
* There should be a connection to the server for any 'new' user log ins. 
* If we are adding something to mobile in the future that needs the full user information, this logic will need to be rewritten.

I still vote for this method, as it's simpler to implement.

Things to consider:
* Every user login, needs to be first checked with the server, and if the server is offline, check against the user table (currently works this way).
* If the user exists in user table, but user authentication against the server fails, then make the user 'inactive' in the user table.
* Now that we will have multiple users, some code needs to be refactored to record currently logged in user in transactions/requisitions (do a global search for `enteredBy` in mobile)

Example of suggested method:
* User table in mobile: (username, password (hashed), active (boolean))
* After initialisation, user logs in, server successfully authenticates the user
* User is save in user table
* User logs out
* User2 tries to log in, server successfully authenticates User2
* User2 is added to user table
* Now, if internet connection is bad, both users can log in offline
* Say, the user log in rights are revoked for that store, next time there in internet connection, and user tries to log in, the server will return failure to authenticate message, and the user is made inactive in mobile user table

Lastly, I do think autocomplete is a good option.

One issues with drop down user list, is users all need to be synced, and if user log in rights are changed on the server, you will need to log in to mobile, then sync and log out, and only then you will see the new user, etc..

**Comment by Chris-Petty** [comment]
@andreievg What you describe is almost exactly how mobile currently works, except currently that you require default store for each user to be the mobile store (which limits the user to one).

**Comment by Chris-Petty** [comment]
> If we change a couple of lines in REST_authenticate method, this issue is resolved..

Essentially yes, but I think syncing the user table will still be beneficial just initially. Maybe restrict it to syncing users that have login rights to the store/site?

> One issues with drop down user list, is users all need to be synced, and if user log in rights are changed on the server, you will need to log in to mobile, then sync and log out, and only then you will see the new user, etc..

 True but a bit of an edge case no? If they are online they should still ping the server rather than the local table. We have the same problem now but we just rarely touch mobile users. That said sync is authenticated with 
```javascript
getAuthHeader() {
    const username = this.settings.get(SYNC_SITE_NAME);
    const password = this.settings.get(SYNC_SITE_PASSWORD_HASH);
    return getAuthHeader(username, password);
}
```
Which are always available, not much reason why we couldn't sync - the actual msupply user's name and pass aren't used for better or worse.

**Comment by sworupSussol** [comment]
***Moving description had with @Chris-Petty here to populate the issue description with relevant information.***

---

(Chris hijack)

Currently only one user is stored on the mobile tablet. Should sync more users so that they can operate offline if they have permission to login (set on primary). This would allow better flexibility for tracking individuals having their own user accounts for entering data.
- Mobile already has a users table, perhaps just on initialisation sync all users so that mobile can remember them, or just sync users that have login rights (former easier and it'll rarely be more than 50 records)
- Review login logic: should still try and ping central server if available, otherwise fallback to local users table.

I think that's all there is to it. Currently you can have multiple users, but their default store has to be the mobile store and I think mobile only remembers the last user that logged in (I could be wrong).

> Can someone explain this issue a bit further please.
> 
> As far as I understood, if two users are trying to connect to a store simultaneously, the system does not allow the second user to log in.

**Comment by sworupSussol** [comment]
@Chris-Petty @andreievg 

I think at this moment what we are trying to do is allow User to login offline after sync, even if they have not logged in with the server at least once. Multi-user login is actually allowed in mobile store in online mode.

In case of mobile, we have only one store attached with a Table so even if we allow one user to log into multiple stores, the user would have to use multiple tables. *Let me know if this needs to be done.*

---

I have done the user table sync and pushed the branches but for now all users will sync to mobile database and since we do not have any access control or enable/disable feature, they will all have log in access. 

Now I have two options:

1. Only sync users that have login access to the store. *I need to change the User table query to exclude users without access to achieve this.*
2. Sync user_store table to mobile too and do authentication based on access given in that table. *This might be an overkill for now.*

I am going to do Option #1 Monday, unless there are any better solutions.

---

Fun Question
---

What if two users have different access settings?

**Comment by andreievg** [comment]
@sworupSussol -> 

> I think at this moment what we are trying to do is allow User to login offline after sync, even if they have not logged in with the server at least once. Multi-user login is actually allowed in mobile store in online mode as we speak.

If this is really the case, then yeah user tables need to be synced.

@Chris-Petty 
> True but a bit of an edge case no? If they are online they should still ping the server rather than the local table. We have the same problem now but we just rarely touch mobile users. That said sync is authenticated with

* During development, we do this all the time (changing user password).
* In production, if user forgets the password, it would need to be changed on the server and synced
* Just saying we have to make sure the synchronisation happens when apps starts up (if it isn't already)

I am not a fan of syncing all users to mobile, especially since there is a risk of device being rooted and password exposed. But if it must happen, these are my recommendations:
* Syncing users that have log-in rights only, maybe too difficult
  * We have to make sure if login right is revoked or introduced, the information should be re-synced and the users should be removed or created (so if we are syncing users, keep sync the same as desktop, ideally desktop should only sync users that are relative to store/site)
  * So, it looks like we need to sync user_store and handle them on mobile (similar to item_store_join)
* Security becomes an issue and we need to store all user passwords on mobile hashed, all with unique salts.
   * The server would need to send the generated hash and the salt (so decrypt password, create a hash with random salt, and send both to mobile). This only needs to be done for mobile sites. There is no way yet, to identify a site as mobile (it should be coming soon though)

@sworupSussol 
> In case of mobile, we have only one store attached with a Table so even if we allow one user to log into multiple stores, the user would have to use multiple tables. Let me know if this needs to be done.

Do you mean multiple stores not multiple tables ? if you mean multiple store, then no, mobile site = one store.

@Chris-Petty 
So it is quite tricky to make off-line user log-in work, we should review the original requirements and see if this is actually required.

The way I see it, there is no immediate need to have `off-line` user log in for users that haven't logged in yet in `on-line` mode. I am not sure why that particular functionality is `high` priority. (Did we have multiple requests for this ?)

If it is required and it is high priority, we need to make it `effort medium` at least

**Comment by sworupSussol** [comment]
> If they are online they should still ping the server rather than the local table.
@Chris-Petty This is how it works in online mode. 
- One thing I noticed was the server login code gets a json back but the revised password is never stored in mobile table, the password would never be updated for offline mode. It should be small fix to remedy it, but this problem does exist in live code.

> Just saying we have to make sure the sync happens when apps starts up.
@andreievg the code that I have added with the branch associated with this issue should sync user tables with the server just like the other tables, right?
- Wouldn't the tables sync up when server has the latest changes?

> So, it looks like we need to sync user_store and handle them on mobile.
I agree.

> Security becomes an issue and we need to store all user passwords on mobile hashed, all with unique salts.
@andreievg We are storing hashed password in mobile. Sever actually has a private public key encryption method that can be reversed if you know the keys. 
There is a salt field in mobile db but it is never populated.

> Do you mean multiple stores not multiple tables ? if you mean multiple store, then no, mobile site = one store.
@andreievg I agree we do not need one user multiple stores. I do not understand what you mean by multiple tables.

My Observation
---

- In mobile store there is multiple user login provided the user logs in once with the server, after which offline login also works.
- The issue with different rights and privileges for different users still exists (in live code) in mobile for us, since once the store gets synced there is no permissions sync and hence in mobile if the data is there, it is accessible to all users. I don't know how much of a problem it is, since we are only syncing limited data and the way mobile is designed, as explained by @ujwalSussol, mobile is never supposed to worry about privileges.
- The store has to be the default and only login store for that particular user. One user is locked to one store. There can be multiple users associated with the same store but the store has to be default store for all of them to work in mobile.

My suggestion
---

- The User table syncing code is already done. The user password is decrypted and hashed in server side and then passed to mobile where it is stored in Realm db's User Table.
- As it stands, the code will sync all Users. 
- I suggest we sync user_store table and take can_login field from there to handle login. Provided these mobile tables would momentarily sync with the server, the mobile user login rights should be synced with server's.

**Comment by andreievg** [comment]
@sworupSussol 
> @andreievg the code that I have added with the branch associated with this issue should sync user tables with the server just like the other tables, right?
> Wouldn't the tables sync up when server has the latest changes?

I've had a quick look at the code, it looks good to me on mobile end, but I don't know how users sync on server end, looking at the user table trigger there is a note (Sync from satellite -> primary). One of the reasons I am a bit anti syncing the users from server, is because the difficulty jump to 'unknown' realm.

I see you've added hashed_password as a seperate field in sync json. Can you make sure that non mobile satellite site doesn't break (because it wouldn't find that field in user table). If it does break, we need to only send hashed_password to site type `mobile`.

> @andreievg We are storing hashed password in mobile. Sever actually has a private public key encryption method that can be reversed if you know the keys.
> There is a salt field in mobile db but it is never populated.

Hashed password by itself is not enough. To be secure we need salt and hash, you should be able to randomly generate a salt, and then use it to hash password, before you send it to mobile or when password is change on mSupply Desktop (with the later, we would need some new fields in user table on mSupply Desktop).

I am not sure why salt was not previously used in mobile, maybe because we didn't have all users synced to mobile.

>> Do you mean multiple stores not multiple tables ? if you mean multiple store, then no, mobile site = one store.

> @andreievg I agree we do not need one user multiple stores. I do not understand what you mean by multiple tables.

You've mentioned multiple tables not me hehe :
> In case of mobile, we have only one store attached with a Table so even if we allow one user to log into multiple stores, the user would have to use multiple tables. Let me know if this needs to be done
So I am not sure what you mean

I do agree, don't worry about user permissions atm.

The code on mSupply Desktop (REST_authenticate) method, will need to authenticate users where the store is not the default store, only need to check if user can log in.


**Concerns**

* I know you've already create user sync on mobile, but i still think the better option would be to add users in mobile when the are authenticated via on-line means
* If sync of users is implemented, it does need to be restricted to only sync relevant users for the give store/site (i.e. in Myanmar RSCS datafile there are 273 users and 34472 user_store join, it's not feasible to make this sync)
* I've just had a little check, and users don't sync to sites form primary (when they are changed)

**Comment by Chris-Petty** [comment]
> * If sync of users is implemented, it does need to be restricted to only sync relevant users for the give store/site (i.e. in Myanmar RSCS datafile there are 273 users and 34472 user_store join, it's not feasible to make this sync)

user_store joins could be those only related to the store on the tablet, followed by only users related to those joins. Pretty easy even in 4D query land.

I'm going to try make a summary comment below, this has turned into a bit of a jungle 😁

**Comment by sworup** [comment]
No Resolved by https://github.com/sussol/msupply/pull/2270

**Comment by joshxg** [comment]
# Testing
- [x] A user can login to a mobile store if it is not their default store
